### PR TITLE
ddl: ban the setting replica for tables in system database (MySQL)

### PR DESCRIPTION
**Pull request by AilinKid**
Signed-off-by: AilinKid <314806019@qq.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/22046

### What is changed and how it works?

What's Changed:
forbidden the set replica behavior for tables in system database(MySQL)

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- ddl: ban the setting replica for system database (MySQL)

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -4360,6 +4360,10 @@ func (d *ddl) AlterTableSetTiFlashReplica(ctx sessionctx.Context, ident ast.Iden
 	if err != nil {
 		return errors.Trace(err)
 	}
+	// Ban setting replica count for tables in MySQL system database.
+	if util.IsSysDB(schema.Name.L) {
+		return errors.Trace(errors.New("[ddl] couldn't set tiflash replica for tables in MySQL system database"))
```
Comment (djshow832) [inline ddl/ddl_api.go:4365]: How about reusing an error number, like `ErrUnsupportedDDLOperation`?

File: ddl/table.go [hunk 2]
```diff
@@ -875,6 +876,11 @@ func (w *worker) onSetTableFlashReplica(t *meta.Meta, job *model.Job) (ver int64
 		return ver, errors.Trace(err)
 	}
 
+	// Ban setting replica count for tables in MySQL system database.
+	if tidb_util.IsSysDB(job.SchemaName) {
```
Comment (djshow832) [inline ddl/table.go:880]: Why do you need to ban it here?

File: util/misc.go [hunk 3]
```diff
@@ -189,6 +189,11 @@ func IsMemDB(dbLowerName string) bool {
 	return false
 }
 
+// IsSysDB checks whether dbLowerName is system database.
+func IsSysDB(dbLowerName string) bool {
```
Comment (djshow832) [inline util/misc.go:193]: So is it allowed to set replicas for `information_schema`?

File: ddl/db_test.go [hunk 4]
```diff
@@ -5295,6 +5295,26 @@ func (s *testSerialDBSuite) TestSetTableFlashReplica(c *C) {
...
 
+func (s *testSerialDBSuite) TestSetTableFlashReplicaForSystemTable(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use mysql")
+	sysTables := make([]string, 0, 24)
+	rows := tk.MustQuery("show tables").Rows()
+	for i := 0; i < len(rows); i++ {
+		sysTables = append(sysTables, rows[i][0].(string))
+	}
+	fmt.Println(sysTables)
```
Comment (djshow832) [inline ddl/db_test.go]:
```suggestion
```

File: util/misc.go (hunk 3 above)
Comment (AilinKid) [reply util/misc.go:193]: Hard to say, from the requirements, MySQL is the only one mentioned.

File: ddl/table.go (hunk 2 above)
Comment (crazycs520) [reply ddl/table.go:880]: Double check?

Comment (AilinKid) [reply ddl/table.go:880]: In the rolling upgrade cases, there must be some old jobs passed and set into the queue already. Banning it here is a kind of double-check.

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -4360,6 +4360,10 @@ func (d *ddl) AlterTableSetTiFlashReplica(ctx sessionctx.Context, ident ast.Iden
 	if err != nil {
 		return errors.Trace(err)
 	}
+	// Ban setting replica count for tables in MySQL system database.
+	if util.IsSysDB(schema.Name.L) {
```
Comment (crazycs520) [inline ddl/ddl_api.go:4364]: should use `IsMemOrSysDB` function?

**Comment by AilinKid** [comment]
@zz-jason should we ban the setting replica for tables in other databases? (for now, from the requirements in Jira issue, we should only ban it in the MySQL database, otherwise, restart TiDB may get into struck)

File: ddl/db_test.go [hunk 6]
```diff
@@ -5299,6 +5299,29 @@ func (s *testSerialDBSuite) TestSetTableFlashReplica(c *C) {
...
+		for i := 0; i < len(rows); i++ {
+			sysTables = append(sysTables, rows[i][0].(string))
+		}
+		for _, one := range sysTables {
+			// Issue: https://github.com/pingcap/tidb/issues/24742
+			// Skip the `stats_extended` system table temporarily.
+			if one == "stats_extended" {
+				continue
+			}
+			_, err := tk.Exec(fmt.Sprintf("alter table %s set tiflash replica 1", one))
```
Comment (djshow832) [inline ddl/db_test.go:5318]:
```suggestion
			_, err := tk.Exec(fmt.Sprintf("alter table `%s` set tiflash replica 1", one))
```

File: ddl/ddl_api.go [hunk 7]
```diff
@@ -4372,6 +4372,10 @@ func (d *ddl) AlterTableSetTiFlashReplica(ctx sessionctx.Context, ident ast.Iden
 	if err != nil {
 		return errors.Trace(err)
 	}
+	// Ban setting replica count for tables in system database.
+	if util.IsMemOrSysDB(schema.Name.L) {
+		return errors.Trace(errUnsupportedAlterReplicaForSysTable)
```
Comment (djshow832) [inline ddl/ddl_api.go:4377]: I think no need to trace it.

File: ddl/table.go [hunk 8]
```diff
@@ -875,6 +876,11 @@ func (w *worker) onSetTableFlashReplica(t *meta.Meta, job *model.Job) (ver int64
 		return ver, errors.Trace(err)
 	}
 
+	// Ban setting replica count for tables in system database.
+	if tidb_util.IsMemOrSysDB(job.SchemaName) {
+		return ver, errors.Trace(errUnsupportedAlterReplicaForSysTable)
```
Comment (djshow832) [inline ddl/table.go:881]: Ditto

File: ddl/db_test.go [hunk 9]
```diff
@@ -5299,6 +5299,24 @@ func (s *testSerialDBSuite) TestSetTableFlashReplica(c *C) {
...
+	memOrSysDB := []string{"MySQL", "INFORMATION_SCHEMA", "PERFORMANCE_SCHEMA", "METRICS_SCHEMA"}
+	for _, db := range memOrSysDB {
+		tk.MustExec("use " + db)
+		rows := tk.MustQuery("show tables").Rows()
+		for i := 0; i < len(rows); i++ {
+			sysTables = append(sysTables, rows[i][0].(string))
+		}
+		for _, one := range sysTables {
+			_, err := tk.Exec(fmt.Sprintf("alter table `%s` set tiflash replica 1", one))
+			c.Assert(err.Error(), Equals, "[ddl:8200]ALTER table replica for tables in system database is currently unsupported")
```
Comment (djshow832) [inline ddl/db_test.go:5314]: Where does the `ddl:8200` come? Will the `8200` change in the future?

Comment (AilinKid) [reply ddl/db_test.go:5314]: it's a common errcode for unsupported ddl ops
`ErrUnsupportedDDLOperation            = 8200`

File: ddl/ddl_api.go (hunk 7 above)
Comment (AilinKid) [reply ddl/ddl_api.go:4377]: why? trace is good for locating the code in the user log in different tidb version.

**Pull request merged**